utils/dvb/dvb-fe-tool.1
.TH "dvb-fe-tool" 1 "Fri Oct 3 2014" "DVBv5 Utils" "User Commands"
.SH NAME
dvb\-fe\-tool \- DVBv5 tool for frontend settings inspect/change
.SH SYNOPSIS
.B dvb\-fe\-tool
[\fIOPTION\fR]...
.SH DESCRIPTION
dvbv5\-fe\-tool is a command line tool for digital TV services that is compliant
with version 5 of the DVB API, and backward compatible with the older
v3 DVB API.
.PP
dvb\-fe\-tool is a tool to inspect and change the tuning parameters directly
at the Digital TV frontend. If called without any parameter, it will show
the device capabilities.
.SH "OPTIONS"
.TP
The following options are valid:
.TP
\fB\-3\fR, \fB\-\-dvbv3\fR
Use DVBv3 only
.TP
\fB\-a\fR, \fB\-\-adapter\fR=\fIADAPTER\fR
Digital TV adapter to use. Default: 0.
.TP
\fB\-A\fR, \fB\-\-acoustical\fR
bips if signal quality is good. Also enables femon mode. Please notice that
console bip should be enabled on your wm.
.TP
\fB\-d\fR, \fB\-\-set\-delsys\fR=\fIPARAMS\fR
Sets delivery system to the one specified at \fIPARAMS\fR. use \fIhelp\fR to
show all supported delivery systems.
.TP
\fB\-f\fR, \fB\-\-frontend\fR=\fIFRONTEND\fR
Digital TV frontend to use. Default: 0.
.TP
\fB\-g\fR, \fB\-\-get\fR
Gets frontend parameters.
.TP
\fB\-m\fR, \fB\-\-femon\fR
Monitors the frontend locking status and the available statistics for a
frontend that it is already being streaming via some other application.
This opens the frontend on read-only mode.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Enables debug messages.
.TP
\fB\-?\fR, \fB\-\-help\fR
Outputs the usage help.
.TP
\fB\-\-usage\fR
Gives a short usage message.
.TP
\fB\-V\fR, \fB\-\-version\fR
Prints program version.
.SH EXIT STATUS
On success, it returns 0.
.SH EXAMPLES
.RS
.SS Showing device capabilities
This is the default behavior. It will open the frontend and show what
capabilities it supports and what's the version of the DVB API.
.PP
.nf
$ \fBdvb\-fe\-tool\fR
Device DRXK DVB\-C DVB\-T (/dev/dvb/adapter0/frontend0) capabilities:
     CAN_FEC_1_2
     CAN_FEC_2_3
     CAN_FEC_3_4
     CAN_FEC_5_6
     CAN_FEC_7_8
     CAN_FEC_AUTO
     CAN_GUARD_INTERVAL_AUTO
     CAN_HIERARCHY_AUTO
     CAN_INVERSION_AUTO
     CAN_MUTE_TS
     CAN_QAM_16
     CAN_QAM_32
     CAN_QAM_64
     CAN_QAM_128
     CAN_QAM_256
     CAN_RECOVER
     CAN_TRANSMISSION_MODE_AUTO
DVB API Version 5.10, Current v5 delivery system: DVBC/ANNEX_A
Supported delivery systems:
    [DVBC/ANNEX_A]
     DVBC/ANNEX_C
     DVBT
.fi
.SS Setting the delivery system
Some devices (like the above example) supports multiple delivery systems.
Older DVB applications might not be able to switch from one delivery system
to another one, as this feature were added only on DVB API version 5.5.
.PP
This feature allows to use those legacy applications.
.PP
.nf
$ \fBdvb\-fe\-tool \-d dvbc/annex_c\fR
Changing delivery system to: DVBC/ANNEX_C
.fi
.SS Getting the current frontend settings
.PP
The tool can also be used to get the current settings. In this case, it will
open the frontend at read-only mode, allowing it to be called while some other
digital TV application is running.
.PP
.nf
$ \fBdvb\-fe\-tool \-g\fR
FREQUENCY = 573000000
MODULATION = QAM/256
INVERSION = AUTO
SYMBOL_RATE = 5217000
INNER_FEC = NONE
DELIVERY_SYSTEM = DVBC/ANNEX_A
.fi
.SS Monitoring the frontend
.PP
The dvb\-fe\-tool can also be used to monitor a DVB frontend statistics
without touching on the device, using the \-\-femon (or \-m) parameter.
The typical usage is to be able to read the statistics while some other
DVB program or application is in use.
.PP
On this mode, it will keep printing the frontend statistics on every second,
until the program is aborted with CTRL\-C:
.PP
For example, while dvbv5\-zap is tuning some channel, it will produce:
.PP
.nf
$ \fBdvb\-fe\-tool \-\-femon\fR
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.40dB UCB= 370 postBER= 28.0x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.80dB UCB= 370 postBER= 25.3x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.30dB UCB= 370 postBER= 25.7x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.30dB UCB= 370 postBER= 28.4x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.50dB UCB= 370 postBER= 26.5x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.50dB UCB= 370 postBER= 26.9x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.20dB UCB= 370 postBER= 28.6x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.70dB UCB= 370 postBER= 26.7x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.80dB UCB= 370 postBER= 23.3x10^-6 PER= 0
Lock   (0x1f) Quality= Good Signal= 100.00% C/N= 36.50dB UCB= 370 postBER= 26.7x10^-6 PER= 0
.fi
.PP
If the standard error device is the console, the messages will be colored
according with the signal quality.
.PP
The presented statistics types and unities will depend on the ones that the
hardware you're using support. So, it may vary from hardware to hardware.
.PP
On this mode, it also supports to output an audible signal if the quality is
good, using the '\-A' parameter.
.PP
Please notice that, on modern Linux systems, the system audio should be
enabled at your window manager and the audio theme should be set to produce
an audio when BELL (\fIcharacter\fP) is sent to the terminal.
.SS NOTE:
C/N on the above stats means Carrier to Noise ratio. This is the Signal to
Noise ratio measured at the pilot carrier or just the Signal to Noise ratio
for modulation types where just one carrier is used, like DVB-C or DVB-S.
.RE
.SH BUGS
Report bugs to \fBLinux Media Mailing List <linux-media@vger.kernel.org>\fR
.SH COPYRIGHT
Copyright (c) 2011\-2014 by Mauro Carvalho Chehab.
.PP
License GPLv2: GNU GPL version 2 <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


ChangeLog
v4l-utils-1.5.91 (Development snapshot)
---------------------------------------

Gregor Jasny (3):
      Start v4l-utils 1.5.91 development cycle
      libdvbv5: Remove left over include of desc_service_list.h
      libdvbv5: Fix typo in doxygen index file

Mauro Carvalho Chehab (30):
      dvb-file: Fix a memory leak
      dvb-file: frees channel/vchannel if service id not found
      dvb-file: Remove spaces at the beginning of channel name
      dvb-file: Only print SDT warning if are there any channel to be stored
      dvb-file: Fix channel name when SDT is not found
      libdvbv5/descriptors: Fix content hexdump
      libdvbv5: Fix the number of frequencies at ISDB-T delivery sys descriptor
      libdvbv5: print ISDB-T frequencies in MHz
      libdvbv5: add a parser for ISO-10646/UCS2 strings
      libdvbv5: Fix ISDB-T TS info descriptor
      dvb-file: Fix no-SDT channel name allocation
      dvb-file: fix the memory deallocation on error
      dvb-file: don't dereference a NULL pointer
      libdvbv5: check for alloc errors at desc_ca
      dvb-format-convert: make valgrind happy
      libdvbv5: fix a nasty to track double-free bug
      dvb-fe-tool: abort nicely on femon mode
      libdvbv5: replace @include by @example
      man: move dvb-scan.1 out of Doxygen
      man: move dvb-fe-tool.1 out of Doxygen
      dvb-format-convert: Fix help message
      man: move dvb-format-convert.1 out of Doxygen
      man: add a utils/dvb/dvbv5-zap.1 page and remove from doxygen
      user man pages: Change report bugs to linux media ML
      Revert "dvb-file: don't dereference a NULL pointer"
      libdvbv5: Fix frequency table alloc size at desc_frequency_list
      .gitignore: Ignore quilt patches/ dir
      dvbv5-scan: add vdr format for the --help parameter
      user man pages: Add version to the man pages


v4l-utils-1.5.90 (Development snapshot)
---------------------------------------

André Roth (7):
      libdvbv5: cleanup logging, some memory checks
      libdvbv5: fix satellite LNBf handling
      libdvbv5: fix dvb_fe_dummy()
      libdvbv5: MPEG ES parser documentation
      libdvbv5: use hyperlinks in doxygen PDF
      libdvbv5: MPEG TS parser documentation
      libdvbv5: MPEG TS parser documentation and cleanups

Gregor Jasny (5):
      Start v4l-utils 1.5.0 development cycle
      buildsystem: Use latest ax_prog_doxygen from GNU Autoconf Archive
      libdvbv5: Add todo file
      libdvbv5: remove service_location descriptor
      libdvbv5: remove service_list descriptor
      doxygen: distribute doc/libdvbv5-index.doc in tarball

Hans Verkuil (9):
      v4l2-compliance: fix buffer state test
      (c)v4l-helpers: zero bytesperline, swap arg order
      qv4l2: the helper now zeroes bytesperline
      cv4l-helpers: add missing ioctls
      v4l2-ctl: fix control support for old kernels.
      v4l-helpers.h: fix control ioctls checks
      qv4l2: init dest format when starting vbi/swradio streaming
      v4l2-compliance: fix valgrind warnings
      qv4l2: fix typo: 576 instead of 575

Hans de Goede (3):
      libv4l2: Fix restoring of original dest-fmt after a VIDIOC_S_DV_TIMING
      libv4l2: Set convert_mmap_frame_size as soon as we've a dest_fmt
      libv4l2: Move alignment of dest_fmt resolution to v4l2_set_src_and_dest_format

Mauro Carvalho Chehab (60):
      libdvbv5: add doxygen tags for header.h
      libdvbv5: add doxygen tags to vct.h
      libdvbv5: add doxygen tags to pat.h
      doxygen_libdvbv5.cfg: add header.h, vct.h and pat.h
      libdvbv5: fix some cut-and-paste errors at documentation
      libdvbv5: add doxygen tags to pmt.h
      libdvbv5: better comment the descriptors
      libdvbv5: fix some issues at PAT documentation
      libdvbv5: add documentation for sdt.h
      libdvbv5: some documentation fixups for sdt.h
      libdvbv5: add documentation for the NIT table parser
      libdvbv5: add documentation for crc32.h
      libdvbv5: Add documentation for EIT DVB parser
      libdvbv5: Add a mainpage to the documentation
      libdvbv5: add documentation for ATSC EIT.
      libdvbv5: don't use ATSC_HEADER() macro
      libdvbv5: don't use DVB_DESC_HEADER() macro
      libdvbv5: add documentation for cat.h and atsc_header.h
      libdvbv5: add support for mgt.h
      libdvbv5 documentation review
      libdvbv5: group documentation into modules
      libdvbv5: add a missing macro to dvb_table group for documentation
      doxygen: add dvbv5-tools to the documentation
      dvb-fe-tool: Fix verbosity and comment unused option
      libdvbv5: fix a typo at ASTRA 1E extended LNBf
      Add support for VDR format (only for output)
      libdvbv5: Add support for opening FE on read only mode
      dvb-fe-tool: add support to open DVB FE in read only mode
      libdvbv5: Don't double-free the file name
      dvb-fe-tool: add femon parameter
      libdvbv5: fix a nasty bug when printing stats
      libdvbv5: Re-add the status name at the stats report
      libdvbv5: export dvb_fe_retrieve_quality()
      dvb-fe-tool: Colorize the status report
      dvb-fe-tool: implement acoustic mode
      update dvb-fe-tool Doxygen documentation
      libdvbv5: revert an ABI breakage
      libdvbv5: Create doxygen docs for mpeg_es.h
      libdvbv5: add documentation for ATSC service location
      libdvbv5: Add documentation for CA descriptor
      libdvbv5: Add documentation for CA identifier descriptor
      libdvbv5: add documentation for cable delivery system descriptor
      libdvbv5: add documentation for the extended event descriptor
      libdvbv5: add documentation for the short event descriptor
      libdvbv5: add documentation for the extended descriptor
      libdvbv5: add documentation for the frequency list descriptor
      libdvbv5: Document the hierarchy descriptor
      libdvbv5: Add documentation for ISDB-T terrestrial descriptor
      libdvbv5: add documentation for the language descriptor
      libdvbv5: Add description for the LCN descriptor

utils/dvb/dvb-fe-tool.c
/*
 * Copyright (c) 2011-2012 - Mauro Carvalho Chehab
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation version 2
 * of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 * Or, point your browser to http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 *
 */

#include "libdvbv5/dvb-file.h"
#include <config.h>
#include <argp.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#define PROGRAM_NAME	"dvb-fe-tool"

const char *argp_program_version = PROGRAM_NAME " version " V4L_UTILS_VERSION;
const char *argp_program_bug_address = "Mauro Carvalho Chehab <m.chehab@samsung.com>";

static const char doc[] = "\nA DVB frontend tool using API version 5\n"
	"\nOn the options bellow, the arguments are:\n"
	"  ADAPTER      - the dvb adapter to control\n"
	"  FRONTEND     - the dvb frontend to control";

static const struct argp_option options[] = {
	{"verbose",	'v',	0,		0,	"enables debug messages", 0},
	{"adapter",	'a',	"ADAPTER",	0,	"dvb adapter", 0},
	{"frontend",	'f',	"FRONTEND",	0,	"dvb frontend", 0},
	{"set-delsys",	'd',	"PARAMS",	0,	"set delivery system", 0},
	{"femon",	'm',	0,		0,	"monitors frontend stats on an streaming frontend", 0},
	{"acoustical",	'A',	0,		0,	"bips if signal quality is good. Also enables femon mode. Please notice that console bip should be enabled on your wm.", 0},
#if 0 /* Currently not implemented */
	{"set",		's',	"PARAMS",	0,	"set frontend", 0},
#endif
	{"get",		'g',	0,		0,	"get frontend", 0},
	{"dvbv3",	'3',	0,		0,	"Use DVBv3 only", 0},
	{ 0, 0, 0, 0, 0, 0 }
};

static int adapter = 0;
static int frontend = 0;
static unsigned get = 0;
static char *set_params = NULL;
static int verbose = 0;
static int dvbv3 = 0;
static int delsys = 0;
static int femon = 0;
static int acoustical = 0;
static int timeout_flag = 0;

static void do_timeout(int x)
{
	(void)x;

	if (timeout_flag == 0) {
		timeout_flag = 1;
		alarm(2);
		signal(SIGALRM, do_timeout);
	} else {
		/* something has gone wrong ... exit */
		exit(1);
	}
}

#define PERROR(x...)                                                    \
	do {                                                            \
		fprintf(stderr, "ERROR: ");                             \
		fprintf(stderr, x);                                     \
		fprintf(stderr, " (%s)\n", strerror(errno));		\
	} while (0)

static error_t parse_opt(int k, char *arg, struct argp_state *state)
{
	switch (k) {
	case 'a':
		adapter = atoi(arg);
		break;
	case 'f':
		frontend = atoi(arg);
		break;
	case 'd':
		delsys = dvb_parse_delsys(arg);
		if (delsys < 0)
			return ARGP_ERR_UNKNOWN;
		break;
	case 'm':
		femon++;
		break;
	case 'A':
		femon++;
		acoustical++;
		break;
#if 0
	case 's':
		set_params = arg;
		break;
#endif
	case 'g':
		get++;
		break;
	case '3':
		dvbv3++;
		break;
	case 'v':
		verbose	++;
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

static struct argp argp = {
	.options = options,
	.parser = parse_opt,
	.doc = doc,
};

static int print_frontend_stats(FILE *fd,
				struct dvb_v5_fe_parms *parms)
{
	char buf[512], *p;
	int rc, i, len, show, n_status_lines = 0;

	rc = dvb_fe_get_stats(parms);
	if (rc) {
		PERROR("dvb_fe_get_stats failed");
		return -1;
	}

	p = buf;
	len = sizeof(buf);
	dvb_fe_snprintf_stat(parms,  DTV_STATUS, NULL, 0, &p, &len, &show);

	for (i = 0; i < MAX_DTV_STATS; i++) {
		show = 1;

		dvb_fe_snprintf_stat(parms, DTV_QUALITY, "Quality",
				     i, &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_STAT_SIGNAL_STRENGTH, "Signal",
				     i, &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_STAT_CNR, "C/N",
				     i, &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_STAT_ERROR_BLOCK_COUNT, "UCB",
				     i,  &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_BER, "postBER",
				     i,  &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_PRE_BER, "preBER",
				     i,  &p, &len, &show);

		dvb_fe_snprintf_stat(parms, DTV_PER, "PER",
				     i,  &p, &len, &show);
		if (p != buf) {
			if (isatty(fileno(fd))) {
				enum dvb_quality qual;
				int color;

				qual = dvb_fe_retrieve_quality(parms, 0);

				switch (qual) {
				case DVB_QUAL_POOR:
					color = 31;
					break;
				case DVB_QUAL_OK:
					color = 36;
					break;
				case DVB_QUAL_GOOD:
					color = 32;
					break;
				case DVB_QUAL_UNKNOWN:
				default:
					color = 0;
					break;
				}
				fprintf(fd, "\033[%dm", color);
				/*
				 * It would be great to change the BELL
				 * tone depending on the quality. The legacy
				 * femon used to to that, but this doesn't
				 * work anymore with modern Linux distros.
				 *
				 * So, just print a bell if quality is good.
				 *
				 * The console audio should be enabled
				 * at the window manater for this to
				 * work.
				 */
				if (acoustical) {
					if (qual == DVB_QUAL_GOOD)
						fprintf(fd, "\a");
				}
			}

			if (n_status_lines)
				fprintf(fd, "\t%s\n", buf);
			else
				fprintf(fd, "%s\n", buf);

			n_status_lines++;

			p = buf;
			len = sizeof(buf);
		}
	}

	fflush(fd);

	return 0;
}

static void get_show_stats(struct dvb_v5_fe_parms *parms)
{
	int rc;

	signal(SIGTERM, do_timeout);
	signal(SIGINT, do_timeout);

	do {
		rc = dvb_fe_get_stats(parms);
		if (!rc)
			print_frontend_stats(stderr, parms);
		if (!timeout_flag)
			usleep(1000000);
	} while (!timeout_flag);
}

int main(int argc, char *argv[])
{
	struct dvb_v5_fe_parms *parms;
	int fe_flags = O_RDWR;

	argp_parse(&argp, argc, argv, 0, 0, 0);

	/*
	 * If called without any option, be verbose, to print the
	 * DVB frontend information.
	 */
	if (!get && !delsys && !set_params && !femon)
		verbose++;

	if (!delsys && !set_params)
		fe_flags = O_RDONLY;

	parms = dvb_fe_open_flags(adapter, frontend, verbose, dvbv3,
				  NULL, fe_flags);
	if (!parms)
		return -1;

	if (delsys) {
		printf("Changing delivery system to: %s\n",
			delivery_system_name[delsys]);
		dvb_set_sys(parms, delsys);
		goto ret;
	}

#if 0
	if (set_params)
		do_something();
#endif
	if (get) {
		dvb_fe_get_parms(parms);
		dvb_fe_prt_parms(parms);
	}

	if (femon)
		get_show_stats(parms);

ret:
	dvb_fe_close(parms);

	return 0;
}
